File: 06. Backtracking/06. Combinations I/main.py
def backtrack(start, n, k, current, result):
    if len(current) == k:
        result.append(list(current))
        return

    for i in range(start, n + 1):
        current.append(i)
        backtrack(i + 1, n, k, current, result)
        current.pop()  # Backtrack

def combine(n, k):
    result = []
    backtrack(1, n, k, [], result)
    return result

# Main function
def main():
    n, k = 4, 2
    result = combine(n, k)

    print("Combinations:")
    print(result)

main()

File: 06. Backtracking/06. Combinations I/main.java
import java.util.*;

public class main {
    public static void backtrack(int start, int n, int k, List<Integer> current, List<List<Integer>> result) {
        if (current.size() == k) {
            result.add(new ArrayList<>(current));
            return;
        }

        for (int i = start; i <= n; i++) {
            current.add(i);
            backtrack(i + 1, n, k, current, result);
            current.remove(current.size() - 1); // Backtrack
        }
    }

    public static List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> result = new ArrayList<>();
        backtrack(1, n, k, new ArrayList<>(), result);
        return result;
    }

    public static void main(String[] args) {
        int n = 4, k = 2;
        List<List<Integer>> result = combine(n, k);

        System.out.println("Combinations:");
        for (List<Integer> comb : result) {
            System.out.println(comb);
        }
    }
}
